rust/src/records.rs

```diff
@@ -1,507 +1,537 @@
-use std::convert::TryFrom;
-
-pub type Timestamp = u64;
-
-#[derive(Debug, Copy, PartialEq, Eq, Serialize, Deserialize, Clone)]
-pub enum OpCode {
-    Header,
-    Footer,
-    Schema,
-    Channel,
-    Message,
-    Chunk,
-    MessageIndex,
-    ChunkIndex,
-    Attachment,
-    AttachmentIndex,
-    Statistics,
-    Metadata,
-    MetadataIndex,
-    SummaryOffset,
-    DataEnd,
-    UserOpcode(u8),
-}
 
-pub struct InvalidOpcode(pub u8);
-
-impl TryFrom<u8> for OpCode {
-    type Error = InvalidOpcode;
-
-    fn try_from(value: u8) -> Result<Self, Self::Error> {
-        match value {
-            0x00 => Err(InvalidOpcode(value)),
-            0x01 => Ok(OpCode::Header),
-            0x02 => Ok(OpCode::Footer),
-            0x03 => Ok(OpCode::Schema),
-            0x04 => Ok(OpCode::Channel),
-            0x05 => Ok(OpCode::Message),
-            0x06 => Ok(OpCode::Chunk),
-            0x07 => Ok(OpCode::MessageIndex),
-            0x08 => Ok(OpCode::ChunkIndex),
-            0x09 => Ok(OpCode::Attachment),
-            0x0A => Ok(OpCode::AttachmentIndex),
-            0x0B => Ok(OpCode::Statistics),
-            0x0C => Ok(OpCode::Metadata),
-            0x0D => Ok(OpCode::MetadataIndex),
-            0x0E => Ok(OpCode::SummaryOffset),
-            0x0F => Ok(OpCode::DataEnd),
-            x if x < 0x80 => Err(InvalidOpcode(x)),
-            x => Ok(OpCode::UserOpcode(x)),
+/// Opcodes for MCAP file records.
+///
+/// "Records are identified by a single-byte opcode.
+/// Record opcodes in the range 0x01-0x7F are reserved for future MCAP format usage.
+/// 0x80-0xFF are reserved for application extensions and user proposals."
+pub mod op {
+    pub const HEADER: u8 = 0x01;
+    pub const FOOTER: u8 = 0x02;
+    pub const SCHEMA: u8 = 0x03;
+    pub const CHANNEL: u8 = 0x04;
+    pub const MESSAGE: u8 = 0x05;
+    pub const CHUNK: u8 = 0x06;
+    pub const MESSAGE_INDEX: u8 = 0x07;
+    pub const CHUNK_INDEX: u8 = 0x08;
+    pub const ATTACHMENT: u8 = 0x09;
+    pub const ATTACHMENT_INDEX: u8 = 0x0A;
+    pub const STATISTICS: u8 = 0x0B;
+    pub const METADATA: u8 = 0x0C;
+    pub const METADATA_INDEX: u8 = 0x0D;
+    pub const SUMMARY_OFFSET: u8 = 0x0E;
+    pub const DATA_END: u8 = 0x0F;
+}
+
+/// A raw record from an MCAP file.
+///
+/// For records with large slices of binary data (schemas, messages, chunks...),
+/// we use a [`CoW`](std::borrow::Cow) that can either borrow directly from the mapped file,
+/// or hold its own buffer if it was decompressed from a chunk.
+#[derive(Debug, Serialize, Deserialize)]
+pub enum Record<'a> {
+    Header(Header),
+    Footer(Footer),
+    Schema {
+        #[serde(flatten)]
+        header: SchemaHeader,
+        data: Cow<'a, [u8]>,
+    },
+    Channel(Channel),
+    Message {
+        #[serde(flatten)]
+        header: MessageHeader,
+        data: Cow<'a, [u8]>,
+    },
+    Chunk {
+        #[serde(flatten)]
+        header: ChunkHeader,
+        data: &'a [u8],
```

Comment:
why no cow here?

Reply:
I think my thought process was to use normal references for types you were unlikely to copy out or pass around. (For example, it wouldn't be unusual to pass Messages, Channels, and Schemas (either from thread-to-thread in a reader, or in the writer API), but you're probably not going to manually assemble a Chunk.

But typing that out makes it sound a bit odd. Want these to be Cow for uniformity's sake?

Reply:
It seems sensible to give the user the option here (and for attachments).

Reply:
Done with 4ded9c7, let me know what you think.